### PR TITLE
Template instantiation

### PR DESCRIPTION
**Pull request by fitzgen**
I very much suspect this needs a rebase, but I haven't done it yet.

This is the majority of #446, although it doesn't have the name generation for anonymous named types patches yet (that will be a different PR).

### CODE REVIEW

**Comment by fitzgen** [comment]
r? @emilio 

Ok, I think this is ready for review. Phew!

File: src/ir/context.rs [hunk 1]
```diff
@@ -1148,6 +1166,68 @@ impl<'ctx> BindgenContext<'ctx> {
...
+impl TemplateDeclaration for PartialType {
+    fn template_params(&self, _ctx: &BindgenContext) -> Option<Vec<ItemId>> {
+        // Maybe at some point we will eagerly parse named types, but for now we
+        // don't and this information is unavailable.
+        None
+    }
+
+    fn num_template_params(&self, _ctx: &BindgenContext) -> Option<usize> {
+        // Wouldn't it be nice if libclang would reliably give us this
+        // information‽
```
Comment (emilio) [inline src/ir/context.rs]: indeed.

**Comment by emilio** [comment]
Oh, and thanks for all the cleanups! :)

**Comment by fitzgen** [comment]
@emilio do you think you could try this out on stylo? I don't have a local set up.

**Comment by fitzgen** [comment]
Thanks!

**Comment by emilio** [comment]
Sure! Any chance to get it rebased? That way I'll also test all the other merged changes at the same time :)

**Comment by fitzgen** [comment]
Working on a rebase

**Comment by fitzgen** [comment]
@emilio rebased!

**Comment by fitzgen** [comment]
This is running into some issues within the `<type_traits>` header, our old nemesis >:-|

**Comment by fitzgen** [comment]
Ok latest commit lets me generate SM bindings (which pulls in `<type_traits>`) again.

**Pull request merged**